internal/cli/task_cancel.go

```diff
@@ -0,0 +1,103 @@
+package cli
+
+import (
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"github.com/hashicorp/waypoint/internal/clierrors"
+	"github.com/hashicorp/waypoint/internal/pkg/flag"
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+type TaskCancelCommand struct {
+	*baseCommand
+
+	flagRunJobId string
+}
+
+func (c *TaskCancelCommand) Run(args []string) int {
+	// Initialize. If we fail, we just exit since Init handles the UI.
+	if err := c.Init(
+		WithArgs(args),
+		WithFlags(c.Flags()),
+	); err != nil {
+		return 1
+	}
+	ctx := c.Ctx
+
+	var taskId string
+	if len(c.args) == 0 && c.flagRunJobId == "" {
+		c.ui.Output("Task ID or Run Job Id required.\n\n%s", c.Help(), terminal.WithErrorStyle())
+		return 1
+	} else {
+		taskId = c.args[0]
+	}
+
+	if c.flagRunJobId != "" && taskId != "" {
+		c.ui.Output("Both Run Job Id and Task Id was supplied, will look up by Task Id", terminal.WithWarningStyle())
+	}
+
+	var (
+		taskReq *pb.CancelTaskRequest
+	)
+
+	if taskId != "" {
+		taskReq = &pb.CancelTaskRequest{
+			Ref: &pb.Ref_Task{
+				Ref: &pb.Ref_Task_Id{
+					Id: taskId,
+				},
+			},
+		}
+	} else if c.flagRunJobId != "" {
+		taskReq = &pb.CancelTaskRequest{
+			Ref: &pb.Ref_Task{
+				Ref: &pb.Ref_Task_JobId{
+					JobId: c.flagRunJobId,
+				},
+			},
+		}
+	}
+
+	_, err := c.project.Client().CancelTask(ctx, taskReq)
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			c.ui.Output("Task not found: %s", clierrors.Humanize(err),
+				terminal.WithErrorStyle())
+			return 1
+		}
+
+		c.ui.Output(clierrors.Humanize(err), terminal.WithErrorStyle())
+		return 1
+	}
+
+	c.ui.Output("Task %q has been requested to be cancelled", taskId, terminal.WithWarningStyle())
+	return 0
+}
+
+func (c *TaskCancelCommand) Flags() *flag.Sets {
+	return c.flagSet(flagSetOperation, func(set *flag.Sets) {
+		f := set.NewSet("Command Options")
+
+		f.StringVar(&flag.StringVar{
+			Name:    "run-job-id",
+			Target:  &c.flagRunJobId,
+			Default: "",
+			Usage:   "Cancel a Task by Run Job Id.",
+		})
+	})
+}
+
+func (c *TaskCancelCommand) Synopsis() string {
+	return "Cancel an On-Demand Runner Task running in Waypoint"
```

Comment:
> an On-Demand
Does this only work with ODR?

Reply:
Yep, you can only cancel tasks with this CLI, and ODR is what spawns them.

Reply:
So I can't cancel a task that's done locally, if I don't have ODR going, or local things don't create tasks?

Reply:
Yeah - ODR in a local context won't work. That would create a single job run via the local runner in the CLI! So local things don't create tasks, only a remote runs.